rita_client/src/operator_update/mod.rs

```diff
@@ -626,65 +596,39 @@ fn contains_forbidden_key(map: Map<String, Value>, forbidden_values: &[&str]) ->
     false
 }
 
-/// Given a vecdeque of usage hours, add up to a month's worth of hours to a returned vecdeque
-pub fn iterate_month_usage_data(mut data: VecDeque<RCUsageHour>) -> VecDeque<UsageHour> {
-    // one month in hours
-    let max_hour_iterations: u32 = 730;
-    let mut client_iter = 0;
-    let mut res = VecDeque::new();
-    while let Some(hour) = data.pop_front() {
-        // either we hit max iterations or we are on the second to last entry.
-        res.push_back(UsageHour {
-            up: hour.up,
-            down: hour.down,
-            price: hour.price,
-            index: hour.index,
-        });
-        if client_iter >= max_hour_iterations || data.len() == 1 {
-            break;
+/// This function handles preparing usage data for upload to operator tools
+fn prepare_usage_data_for_upload(
+    ops_last_seen_hour: Option<u64>,
+) -> Result<Option<UsageTrackerTransfer>, RitaClientError> {
+    // if this is none we will not send usage data to ops
+    // as it indicates we don't yet know what data ops has
+    match ops_last_seen_hour {
+        Some(ops_last_seen_hour) => {
+            let current_hour = get_current_hour()?;
+            let usage_data_client = get_usage_data_map(Client);
+            let usage_data_relay = get_usage_data_map(Relay);
+            // We check that the difference is >1 because we leave a 1 hour buffer to prevent from sending over an incomplete current hour
+            let min_send_hour = current_hour - ops_last_seen_hour;
+            let mut client_bandwidth = HashMap::new();
+            let mut relay_bandwidth = HashMap::new();
+            for (index, usage) in usage_data_client {
+                if index > min_send_hour {
```

Comment:
You're right this logic is incorrect. I intended min_send_hour to be the last seen hour, thanks for the flag.

As far as looping O(n) is the best possible in this case, because keys in a hashmap are unordered, we could get the keys, sort them, then request only keys greater than last seen hour. But that sorting operation would be O(n)log(n) so just iterating over everything once is actually slightly faster. At the scale of data we're working at an O(n) iteration over data in memory is pretty trivial as far as computation time goes especially considering it will be a few years worth of data at most, that's just a few thousand hours.